pkg/controller/vitessshard/reconcile_rollout.go

```diff
@@ -0,0 +1,214 @@
+package vitessshard
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apilabels "k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	planetscalev2 "planetscale.dev/vitess-operator/pkg/apis/planetscale/v2"
+	"planetscale.dev/vitess-operator/pkg/operator/results"
+	"planetscale.dev/vitess-operator/pkg/operator/rollout"
+)
+
+func (r *ReconcileVitessShard) reconcileRollout(ctx context.Context, vts *planetscalev2.VitessShard) (reconcile.Result, error) {
+	resultBuilder := &results.Builder{}
+
+	if !rollout.Scheduled(vts) {
+		// If the shard is not scheduled for a rolling update, silently bail out and do nothing.
+		return resultBuilder.Result()
+	}
+
+	for _, tablet := range vts.Status.Tablets {
+		if tablet.Ready != corev1.ConditionTrue {
+			if tablet.Ready != corev1.ConditionTrue {
+				// If all tablets aren't healthy, we should bail and not perform a rolling restart.
+				r.recorder.Eventf(vts, corev1.EventTypeWarning, "RollingRestartFailed", "all tablets are not healthy")
+				return resultBuilder.Result()
+			}
+		}
+	}
+
+	podList := &corev1.PodList{}
+	listOpts := &client.ListOptions{
+		Namespace: vts.Namespace,
+		LabelSelector: apilabels.Set(map[string]string{
+			planetscalev2.ComponentLabel: planetscalev2.VttabletComponentName,
+			planetscalev2.ClusterLabel:   vts.Labels[planetscalev2.ClusterLabel],
+			planetscalev2.KeyspaceLabel:  vts.Labels[planetscalev2.KeyspaceLabel],
+			planetscalev2.ShardLabel:     vts.Spec.KeyRange.SafeName(),
+		}).AsSelector(),
+	}
+	if err := r.client.List(ctx, listOpts, podList); err != nil {
+		return resultBuilder.Error(err)
+	}
+
+	scheduledTablets := false
+	for i := range podList.Items {
+		pod := &podList.Items[i]
+		if !rollout.Scheduled(pod) {
+			continue
+		}
+
+		scheduledTablets = true
+		if err := r.releaseTabletPod(ctx, log, pod); err != nil {
+			r.recorder.Eventf(vts, corev1.EventTypeWarning, "RollingRestartFailed", "tablet deletion failed")
+			return resultBuilder.Error(err)
+		}
+	}
+
+	// If we have no more scheduled tablets, unschedule the shard.
+	if !scheduledTablets {
+		if err := r.unscheduleObject(ctx, vts); err != nil {
+			return resultBuilder.Error(err)
+		}
+	}
+
+	return resultBuilder.Result()
+}
+
+func (r *ReconcileVitessShard) releaseTabletPod(ctx context.Context, log *logrus.Entry, pod *corev1.Pod) error {
+	// Release the pod to be recreated with updates.
+	if err := r.releaseObject(ctx,pod); err != nil {
+		return err
+	}
+
+	// TODO: Evict pods before deleting them.
```

Comment:
Eviction is an alternative to deletion; you wouldn't do both.

```suggestion
	// TODO: Evict Pods instead of deleting them directly, to respect PDBs.
```